### PR TITLE
Consolidate infrastructure – adopt `pyproject.toml`

### LINKED ISSUE #40: add `pyproject.toml`, etc.
xref https://github.com/pysal/submodule_template/pull/39#issuecomment-1419446231

We should probably add in `pyproject.toml`, and with that we can streamline:
* `setup.py`, `setup.cfg`
* move requirements into `pyproject.toml` and remove individual reqs files

Moreover, if we want to adopt [`setuptools_scm`](https://github.com/pypa/setuptools_scm) for version control `[ruff](https://github.com/charliermarsh/ruff)` for linting we can:
* get rid of `setup.py`, `setup.cfg`, `versioneer.py`, `_version.py`, `.pylintrc`, `.coveragerc`

cc @knaaptime, @weikang9009, @martinfleis

### PR DESCRIPTION
**Pull request by jGaboardi**
This PR resolves #40.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    "pytest-cov",
+    "pytest-xdist",
+]
+plus = [  # this section can have any name desired; there can be any number of them
+    "ANY_OTHER_OPTIONAL_PACKAGES"
+]
+
+[tool.setuptools.packages.find]
+include = [
+    "spaghetti",
```
Comment (knaaptime) [inline pyproject.toml:71]: PACKAGE_NAME

File: .github/release.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    authors:
+      - dependabot
+  categories:
+    - title: Bug Fixes
+      labels:
+        - bug
+    - title: Enhancements
+      labels:
+        - enhancement
+    - title: LABEL_TITLE_1
```
Comment (knaaptime) [inline .github/release.yml:14]: major? or is that implied with no label?

**Review by knaaptime — COMMENTED** [review]
sweet, thanks

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,97 @@
+[build-system]
+requires = [
+    "black",
+    "isort",
+    "pyproject-flake8",
```
Comment (martinfleis) [inline pyproject.toml:5]: Shouldn't this be replaced by ruff?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+]
+docs = [
+    "nbsphinx",
+    "numpydoc",
+    "sphinx",
+    "sphinxcontrib-bibtex",
+    "sphinx_bootstrap_theme",
+]
+tests = [
+    "codecov"
```
Comment (martinfleis) [inline pyproject.toml:59]:
```suggestion
    "codecov",
```

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+omit = ["PACKAGE_NAME/test_*"]
+source = ["PACKAGE_NAME"]
+
+[tool.coverage.report]
+ignore_errors = true
+omit = ["PACKAGE_NAME/test_*"]
+
+[tool.ruff]
+line-length = 88
+select = ["E", "F", "W", "I", "UP", "N", "B", "A", "C4", "SIM", "ARG"]
```
Comment (martinfleis) [inline pyproject.toml:85]: Not sure if having all these is not too strict. Can we make a note on which are required (F, I, U ??? not sure) and which optional?

**Review by martinfleis — COMMENTED** [review]
Thanks! I like the direction.

File: pyproject.toml (hunk 3 above)
Comment (jGaboardi) [reply pyproject.toml:5]: Yes, my oversight. Was pulling a late night.

File: pyproject.toml (hunk 1 above)
Comment (jGaboardi) [reply pyproject.toml:71]: I made several easy mistakes in this PR... LOL

File: .github/release.yml (hunk 2 above)
Comment (jGaboardi) [reply .github/release.yml:14]: We (or other maintainers) can add any label that is desired so long as we are staying on top of labelling issue and PRs consistently.

File: pyproject.toml (hunk 5 above)
Comment (jGaboardi) [reply pyproject.toml:85]: Yes, I think this is a good idea. I simply copied your's over from [`clustergram`](https://github.com/martinfleis/clustergram/blob/main/pyproject.toml#L53). I like have all of them, but (like you say) they are not all required. Maybe also adding in a comment pointing to [`ruff`'s rule codes](https://beta.ruff.rs/docs/rules/).

Comment (jGaboardi) [reply pyproject.toml]: OK, so I've label these with required, optional (strongly encouraged), and optional. Let me know how you think about that.

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,16 +1,14 @@
...
-  hooks:
-  - id: black
-    language_version: python3
+  - repo: https://github.com/psf/black
+    rev: 22.12.0
+    hooks:
+      - id: black
+        language_version: python3
+  - repo: https://github.com/charliermarsh/ruff-pre-commit
+    rev: "v0.0.230"
```
Comment (martinfleis) [inline .pre-commit-config.yaml:9]:
```suggestion
    rev: "v0.0.251"
```

that is the latest as of today but they have a rapid release cycle

**Review by martinfleis — APPROVED** [review]
one minor comment, otherwise looks good to me. thanks!

**Pull request merged**